Update the code to support Python 3.12.4

JimZhouZZY opened the pull request:
Changes:
1. Replaced the use of the `pkg_resources` method with the newer `importlib` method. For detailed guidance on this, please refer to [this page](https://importlib-resources.readthedocs.io/en/latest/migration.html#pkg-resources-resource-filename).

2. Renamed the file types.py to type_definitions.py to prevent conflicts with the similarly named file in the Python library.

JimZhouZZY commented:
Could not pass the tests due to python version changes.

Grimmys commented:
It's good changes, for a second I thought we were in the RPG project and I was going to say it's maybe to early to do such change (since 3.11 is out since October 24, 2022 only), but it's the library repo so I think it's fine!

Just a few other places should be updated in order to make this upgrade:

- `setup.py` to indicate Python 3.12 is required to build the project
- `.github/workflows/test-library.yml` to change the version of Python used to run the tests, hopefully fixing them

Grimmys commented on src/pygamepopup/configuration.py:
```diff
@@ -8,7 +8,7 @@
 from typing import Union
 
-import pkg_resources
+import importlib.resources as res
```
I don't really like extensive abbreviations, because they can be confusing sometimes.
I suggest to replace it by `from importlib import resources`.

JimZhouZZY replied:
Changed.

JimZhouZZY commented:
Changed import method . 
Modified some other files to adapt the use of importlib. 
Renamed import codes to adapt the new name `type_definitions.py` of the old `types.py`

JimZhouZZY commented:
> It's good changes, for a second I thought we were in the RPG project and I was going to say it's maybe to early to do such change (since 3.11 is out since October 24, 2022 only), but it's the library repo so I think it's fine!
> 
> Just a few other places should be updated in order to make this upgrade:
> 
> * `setup.py` to indicate Python 3.12 is required to build the project
> * `.github/workflows/test-library.yml` to change the version of Python used to run the tests, hopefully fixing them

Thank you for the help, I now successfully passed the unit test. However it failed to build docs. I've checked the error messages, seems like the `__name__` here refers to `pygamepopup.configuration` itself, causing the error.

I would like to use `__package__` instead of `__name__` here to prevent this issue.

Additionaly, I made changed the version number to 0.10.1 and updated CHANGELOG.txt. I'm still not sure about the version number. Please take a look and correct the version number and the changelog if they are not good.

Grimmys commented:
The version should rather be a minor one than a patch one, because it's changing the way consumers should use the module (they cannot integrate it anymore in a project running on Python < 3.12).

I'm still thinking about this change because I didn't immediately realized it would not be backward compatible anymore... Meaning the minimal version of Python for the RPG project for example would have to be increased to 3.12. And 3.12 is a very recent version (less than a year) so not sure if it's fair to force users to be on this latest version?
At the mean time I guess it's good motivation for people to upgrade to this newest version.

JimZhouZZY commented:
Understood, I'll make changes on these version numbers. Let's leave this PR open and merge it later.

Grimmys commented:
I'm merging it.
